### PR TITLE
test(content-style): Expanded CI data file check

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature:** This is a re-PR of #7274, with an entirely new checker script.

## Feature Details
This PR introduces a new script in the `utils` directory: `check_content_style.py`. ~~Wow, great work, right?~~

This script can detect, and - wait for it - automatically correct several formatting issues in data files. It also comes with an intuitive terminal interface, since I expect people to use this script manually, as well as via automated tools.

## Usage Examples
Older versions of this script (mainly what was in #7274) were used in many of my typo fix PRs. This new script is more powerful, more flexible, and more intuitive. At the time of writing, it found 5 'errors' and  20 'warnings' in the data files, even though I had a typo fix pr literally yesterday. (Please note that these 'errors' and 'warnings' are different than what the game's parser produces.)

## Features
- [x] Detects and fixes trailing spaces
- [x] Detects and fixes some non-ASCII quotes and apostrophes
- [x] Detects other non-ASCII characters
- [x] Checks copyright headers
- [x] Detects and fixes duplicate spaces
- [x] Detects and fixes extra spaces when indenting in a conversation
- [x] Detects and fixes missing spaces before comments
- [x] Detects and fixes extra spaces inside quotes at the beginning/end of text
- [x] Detects and fixes incorrect indentation levels (limited)
- [x] Detects and fixes line separators
- [x] Detects and fixes trailing empty lines
- [x] Detects and fixes missing/extra spaces around punctuation (works even with `word`s and substitutions)
- [x] Detects and fixes some split words (an other -> another, space ship -> spaceship); not sure if we should have this, might be removed later.

And all of this can be configured in the newly added `contentStyle.json` file. More information is available by executing the script with the `--help` and `--config-help` options.

The output format of this script is similar to that of the code style checker, except this one also outputs information about corrected errors.

## Discussion
In the old PR, there were concerns that the script wasn't flexible enough and that it would set an unnecessary precedent for plugins - the parser is flexible, after all. This is why this script comes with a full config file, that controls just about every check. It is even possible to pass a config file and change the data directories from the command line, so I think this issue is dealt with.

There were a bunch of other related discussions about what the formatting rules should be for this repo. One of the main problems was the indentation of empty lines. There are currently around 3000 empty lines with indentation, and around 900 without, however there are many people advocating for both side. The script currently supports both, and can even ignore all but the most outrageous issues in this regard, however it is something that we could resolve with this pr.

<details>
<summary>Why the previous PR was closed</summary>

image

</details>

## Testing Done
I ran the script on the data files, which are over 180k lines long. It seems to work.

### Automated Tests Added
A new CI job is added for this script.

## Performance Impact
This script runs within 10 seconds with the default options. It takes a bit longer with auto-correct turned on, but even that tends to fit within 10-15 seconds.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -258,3 +258,21 @@ jobs:
...
+    if: ${{ needs.changed.outputs.data == 'true' }}
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v3
+    - name: set up python environment
+      uses: actions/setup-python@v4
+      with:
+        python-version: '3.x'
+    - name: install python packages
```
Comment (quyykk) [inline .github/workflows/ci.yml:275]:
```suggestion
    - name: Install dependencies
```

**Review by quyykk — COMMENTED** [review]
~~I'm not a fan of the config file~~ at least put it somewhere else, say in utils/

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -258,3 +258,21 @@ jobs:
...
       run: python ./utils/check_code_style.py
+
+
+  check-content-style:
+    name: Content Style Check
+    needs: changed
+    if: ${{ needs.changed.outputs.data == 'true' }}
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply .github/workflows/ci.yml:269]: I literally copy-pasted this job from the code style checker.

**Comment by tibetiroka** [comment]
We have the editorconfig, codespell etc. files in the parent directory, why should we move this to `utils`?

File: .github/workflows/ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (quyykk) [reply .github/workflows/ci.yml:269]: Yes and I don't like it there either :smile:

[the remark this answers is not in the document]
Comment (tibetiroka) [reply .github/workflows/ci.yml:269]: Should I fix both in this PR then, or only this one?

File: contentStyle.json [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+						"description": "missing space after punctuation",
+						"regex": "([^\\s]*[:,.?!;]+)(?!(?:[ \"'`\\)\\]:,.?!;<]|$))([^\\s\"`'.!?]*)",
+						"except": ["[.,][0-9]", "https?://", "[A-Z]\\.", "\\.{3}[\\w\\(\\[\"'`]", "[><]\\?=", "^[\\(\\[]?([\\w-]+\\.)*[\\w-]*[\\)\\]]?$", "^[\"`']*!"],
+						"correction": {
+							"replaceWith": "\\1 \\2"
+						}
+					},
+					{
+						"description": "unnecessarily split word (space ship -> spaceship, an other -> another)",
+						"regex": "(?<![^\\s])(space|an) (ship|other)(?![^\\s])",
```
Comment (petervdmeer) [inline contentStyle.json:99]: This also matches the line "When in space other ships start approaching you". (Which shouldn't be rejected on a split word check.)

**Review by petervdmeer — APPROVED** [review]
One minor comment, and you need to process quyykks review feedback...

File: contentStyle.json (hunk 3 above)
Comment (tibetiroka) [reply contentStyle.json:99]: Yes, it does match. If you check the next line however, you can see that it excludes anything other than 'space ship' or 'an other'. I know it feels convoluted to check for them twice, but this is the most expandable pattern I could find (it is really easy to add other two-part entries without creating a whole new check).

At least, it is supposed to work this way. Did you get a false positive?

**Comment by tibetiroka** [comment]
> you need to process quyykks review feedback

I take it you also want the config file moved to `utils`?

File: contentStyle.json (hunk 3 above)
Comment (petervdmeer) [reply contentStyle.json]: No I didn't get a false positive. This is okay this way.

**Review by petervdmeer — APPROVED** [review]

**Pull request merged**